src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,89 @@
+using GraphQL.Server.Common;
+using System;
+using System.Buffers;
+using System.IO;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace GraphQL.Server.Serialization.SystemTextJson
+{
+    public class GraphQLRequestDeserializer : IGraphQLRequestDeserializer
+    {
+        private readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions();
+
+        public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
+        {
+            configure(_serializerOptions);
+        }
+
+        public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream, long? contentLength)
+        {
+            int length;
+            byte[] jsonBytes;
+            var sharedArrayPool = ArrayPool<byte>.Shared;
+            bool wasRented = false;
+            if (contentLength.HasValue)
+            {
+                length = (int)contentLength;
+                jsonBytes = sharedArrayPool.Rent(length);
+                wasRented = true;
+                await stream.ReadAsync(jsonBytes, 0, length);
+            }
+            else
+            {
+                using var ms = new MemoryStream();
+                await stream.CopyToAsync(ms).ConfigureAwait(false);
+                jsonBytes = ms.ToArray();
+                length = jsonBytes.Length;
+            }
+            try
+            {
+                return Process(jsonBytes, length);
+            }
+            finally
+            {
+                if (wasRented)
+                {
+                    sharedArrayPool.Return(jsonBytes);
+                }
+            }
+        }
+
+        private GraphQLRequestDeserializationResult Process(byte[] jsonBytes, int length)
+        {
+            var jsonData = jsonBytes.AsSpan(0, length);
```

Comment:
AsSpan here is because the ArrayPool can give me a buffer bigger than what I actually needed, so I'm trimming it here.

Reply:
Have added a comment here in latest to explain this better.